snorkel/labeling/model/tuner.py

```diff
@@ -0,0 +1,157 @@
+
+        class LabelModelTrainable(Trainable):
+            """Trainable class for LabelModel for Ray tuner."""
+
+            def _setup(self, config: Optional[Dict[str, Any]]):
+                self.tuner = LabelModelTrainable.tuner  # type: ignore
+                self.tuner.config = config
+                self.timestep = 0
+
+            def _train(self):
+                self.timestep += 1
+                self.tuner.label_model.fit(
+                    self.tuner.L_train, **self.tuner.fit_config, **self.config
+                )
+                score = self.tuner.label_model.score(
+                    self.tuner.L_dev, self.tuner.Y_dev, metrics=[self.tuner.metric]
+                )
+                return score
+
+            def _save(self, checkpoint_dir: str):
+                path = os.path.join(checkpoint_dir, "checkpoint")
+                with open(path, "w") as f:
+                    f.write(json.dumps({"timestep": self.timestep}))
+                return path
+
+            def _restore(self, checkpoint_path: str):
+                with open(checkpoint_path) as f:
+                    self.timestep = json.loads(f.read())["timestep"]
+
+        # returning a LabelModelTrainable class definition with tuner as user-defined LabelModelTuner instance
+        LabelModelTrainable.tuner = self
+        return LabelModelTrainable
+
+    def run_tuner(
+        self,
+        config: Dict[str, Any],
+        scheduler: Optional[Any] = None,
+        name: Optional[str] = "label_model_tuner",
+        num_samples: Optional[int] = 5,
+        verbose: Optional[bool] = True,
+        experiment_configs: Optional[Dict[str, Any]] = {
+            "stop": {"training_iteration": 100}
+        },
+        run_configs: Optional[Dict[str, Any]] = {
+            "raise_on_failed_trial": False,
+            "verbose": False,
+        },
+    ) -> Tuple[Dict[str, float], Dict[str, float]]:
+        """
+        Run Ray tuner with specified scheduler, or random/grid search according to config definition.
+
+        Example
+        -------
+        >>> L_train = np.array([[1, 1, 1, 0], [0, 1, 1, 0], [1, 0, 0, 1], [0, 1, 0, 1]])
+        >>> L_dev = np.array([[1, 1, 1, 0], [0, 1, 1, 0], [1, 0, 0, 1], [0, 1, 0, 1]])
+        >>> Y_dev = np.array([1, 1, 0, 1])
+        >>> label_model = LabelModel(cardinality=2)
+        >>> label_model_tuner = LabelModelTuner(label_model, L_train, L_dev, Y_dev)
+        >>> config = {"lr": ray.tune.grid_search([1e-4]), "l2": ray.tune.grid_search([0, 1])}
+        >>> label_model_tuner.run_tuner(config, num_samples=1)  # doctest: +SKIP
+
+        Parameters
+        ----------
+        config
+            Dictionary of hyperparameters for LabelModel fit() to search over
+        scheduler
+            Special scheduler for hyperparameter search trials
+        name
+            Name of experiment
+        num_samples
+            Number of trials to run. Set to 1 for gridsearch
```

Comment:
From my reading of this, 1 would mean pick one random set of hyperparams and just run that (i.e. not a search). is it a flag value?

Reply:
There's no flag value iirc, but with grid search, one "sample" goes through all the possible combinations by default.

I can change default for num_smaples to be 1 and then set it explicitly for the rest